File: apps/demo-be-csharp-aspnetcore/tests/DemoBeCsas.Tests/Integration/Steps/AttachmentSteps.cs
using System.Text;
using System.Text.Json;
using DemoBeCsas.Tests.ScenarioContext;
using FluentAssertions;
using Reqnroll;
using Xunit;

namespace DemoBeCsas.Tests.Integration.Steps;

[Binding]
[Trait("Category", "Integration")]
public class AttachmentSteps(
    ServiceLayer svc,
    SharedState state,
    ExpenseSteps expenseSteps
)
{
    // ─────────────────────────────────────────────────────────────
    // Given helpers
    // ─────────────────────────────────────────────────────────────

    [Given(@"^alice has uploaded file ""([^""]+)"" with content type ""([^""]+)"" to the entry$")]
    public async Task GivenAliceUploadedFile(string filename, string contentType)
    {
        state.LastExpenseId.Should().NotBeNull("expense ID should be set");
        var data = Encoding.UTF8.GetBytes($"dummy content for {filename}");
        var response = await svc.UploadAttachmentAsync(
            state.AccessToken,
            state.LastExpenseId!.Value,
            filename,
            contentType,
            data
        );
        ((int)response.StatusCode).Should().Be(
            201,
            $"Failed to upload attachment: {response.Body}"
        );
        using var doc = JsonDocument.Parse(response.Body);
        if (doc.RootElement.TryGetProperty("id", out var idEl))
        {
            state.LastAttachmentId = Guid.Parse(idEl.GetString()!);
        }
    }

    // ─────────────────────────────────────────────────────────────
    // When steps
    // ─────────────────────────────────────────────────────────────

    [When(@"^alice uploads file ""([^""]+)"" with content type ""([^""]+)"" to POST /api/v1/expenses/\{expenseId\}/attachments$")]
    public async Task WhenAliceUploadsFileToExpense(string filename, string contentType)
    {
        state.LastExpenseId.Should().NotBeNull("expense ID should be set");
        var data = Encoding.UTF8.GetBytes($"dummy content for {filename}");
        state.LastResponse = await svc.UploadAttachmentAsync(
            state.AccessToken,
            state.LastExpenseId!.Value,
            filename,
            contentType,
            data
        );
        if (state.LastResponse.IsSuccess)
        {
            using var doc = JsonDocument.Parse(state.LastResponse.Body);
            if (doc.RootElement.TryGetProperty("id", out var idEl))
            {
                state.LastAttachmentId = Guid.Parse(idEl.GetString()!);
            }
        }
    }

    [When(@"^alice uploads file ""([^""]+)"" with content type ""([^""]+)"" to POST /api/v1/expenses/\{bobExpenseId\}/attachments$")]
    public async Task WhenAliceUploadsFileToBobsExpense(string filename, string contentType)
    {
        expenseSteps._bobExpenseId.Should().NotBeNull("bob's expense ID should be set");
        var data = Encoding.UTF8.GetBytes($"dummy content for {filename}");
        state.LastResponse = await svc.UploadAttachmentAsync(
            state.AccessToken,
            expenseSteps._bobExpenseId!.Value,
            filename,
            contentType,
            data
        );
    }

    [When(@"^alice uploads an oversized file to POST /api/v1/expenses/\{expenseId\}/attachments$")]
    public async Task WhenAliceUploadsOversizedFile()
    {
        state.LastExpenseId.Should().NotBeNull("expense ID should be set");
        // 11 MB exceeds the 10 MB limit
        var largeData = new byte[11 * 1024 * 1024];
        Array.Fill(largeData, (byte)'x');
        state.LastResponse = await svc.UploadAttachmentAsync(
            state.AccessToken,
            state.LastExpenseId!.Value,
            "large.jpg",
            "image/jpeg",
            largeData
        );
    }

    [When(@"^alice sends GET /api/v1/expenses/\{expenseId\}/attachments$")]
    public async Task WhenAliceListsAttachments()
    {
        state.LastExpenseId.Should().NotBeNull("expense ID should be set");
        state.LastResponse = await svc.ListAttachmentsAsync(
            state.AccessToken,
            state.LastExpenseId!.Value
        );
    }

    [When(@"^alice sends GET /api/v1/expenses/\{bobExpenseId\}/attachments$")]
    public async Task WhenAliceListsBobsAttachments()
    {
        expenseSteps._bobExpenseId.Should().NotBeNull("bob's expense ID should be set");
        state.LastResponse = await svc.ListAttachmentsAsync(
            state.AccessToken,
            expenseSteps._bobExpenseId!.Value
        );
    }

    [When(@"^alice sends DELETE /api/v1/expenses/\{expenseId\}/attachments/\{attachmentId\}$")]
    public async Task WhenAliceDeletesAttachment()
    {
        state.LastExpenseId.Should().NotBeNull("expense ID should be set");
        state.LastAttachmentId.Should().NotBeNull("attachment ID should be set");
        state.LastResponse = await svc.DeleteAttachmentAsync(
            state.AccessToken,
            state.LastExpenseId!.Value,
            state.LastAttachmentId!.Value
        );
    }

    [When(@"^alice sends DELETE /api/v1/expenses/\{bobExpenseId\}/attachments/\{attachmentId\}$")]
    public async Task WhenAliceDeletesBobsAttachment()
    {
        expenseSteps._bobExpenseId.Should().NotBeNull("bob's expense ID should be set");
        state.LastAttachmentId.Should().NotBeNull("attachment ID should be set");
        state.LastResponse = await svc.DeleteAttachmentAsync(
            state.AccessToken,
            expenseSteps._bobExpenseId!.Value,
            state.LastAttachmentId!.Value
        );
    }

    [When(@"^alice sends DELETE /api/v1/expenses/\{expenseId\}/attachments/\{randomAttachmentId\}$")]
    public async Task WhenAliceDeletesNonExistentAttachment()
    {
        state.LastExpenseId.Should().NotBeNull("expense ID should be set");
        var randomId = Guid.NewGuid();
        state.LastResponse = await svc.DeleteAttachmentAsync(
            state.AccessToken,
            state.LastExpenseId!.Value,
            randomId
        );
    }

    // ─────────────────────────────────────────────────────────────
    // Then steps
    // ─────────────────────────────────────────────────────────────

    [Then(@"^the response body should contain (\d+) items in the ""([^""]+)"" array$")]
    public void ThenResponseContainsItemsInArray(int count, string arrayField)
    {
        state.LastResponse.Should().NotBeNull();
        var body = state.LastResponse!.Body;
        using var doc = JsonDocument.Parse(body);
        doc.RootElement.TryGetProperty(arrayField, out var arr)
            .Should().BeTrue($"'{arrayField}' not found in: {body}");
        arr.ValueKind.Should().Be(JsonValueKind.Array);
        arr.GetArrayLength().Should().Be(count);
    }

    [Then(@"^the response body should contain an attachment with ""filename"" equal to ""([^""]+)""$")]
    public void ThenResponseContainsAttachmentWithFilename(string filename)
    {
        state.LastResponse.Should().NotBeNull();
        var body = state.LastResponse!.Body;
        body.Should().Contain(filename, $"Expected attachment with filename '{filename}' in: {body}");
    }

    [Then(@"^the response body should contain an error message about file size$")]
    public void ThenErrorAboutFileSize()
    {
        state.LastResponse.Should().NotBeNull();
        var body = state.LastResponse!.Body;
        // 413 status is sufficient
        body.Should().ContainAny("size", "large", "limit", "413", "exceed");
    }
}

File: apps/demo-be-java-vertx/src/test/java/com/organiclever/demojavx/integration/steps/AttachmentSteps.java
package com.organiclever.demojavx.integration.steps;

import com.organiclever.demojavx.support.AppFactory;
import com.organiclever.demojavx.support.ScenarioState;
import com.organiclever.demojavx.support.ServiceResponse;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import org.junit.jupiter.api.Assertions;

public class AttachmentSteps {

    private static final int MAX_FILE_SIZE = 10 * 1024 * 1024; // 10MB

    private final ScenarioState state;

    public AttachmentSteps(ScenarioState state) {
        this.state = state;
    }

    @When("^alice uploads file \"([^\"]*)\" with content type \"([^\"]*)\" to POST /api/v1/expenses/\\{expenseId\\}/attachments$")
    public void aliceUploadsFileToExpense(String filename, String contentType) throws Exception {
        String token = state.getAccessToken();
        String expenseId = state.getExpenseId();
        Assertions.assertNotNull(expenseId, "Expense ID must be set");

        byte[] content = createSampleContent(filename);
        ServiceResponse response = AppFactory.getService()
                .uploadAttachment(token, expenseId, filename, contentType, content);
        state.setLastResponse(response);
        if (response.statusCode() == 201) {
            JsonObject body = response.body();
            if (body != null) {
                state.setAttachmentId(body.getString("id"));
            }
        }
    }

    @When("^alice uploads file \"([^\"]*)\" with content type \"([^\"]*)\" to POST /api/v1/expenses/\\{bobExpenseId\\}/attachments$")
    public void aliceUploadsFileToBobsExpense(String filename,
            String contentType) throws Exception {
        String token = state.getAccessToken();
        String expenseId = state.getBobExpenseId();
        Assertions.assertNotNull(expenseId, "Bob's expense ID must be set");

        byte[] content = createSampleContent(filename);
        ServiceResponse response = AppFactory.getService()
                .uploadAttachment(token, expenseId, filename, contentType, content);
        state.setLastResponse(response);
    }

    @When("^alice uploads an oversized file to POST /api/v1/expenses/\\{expenseId\\}/attachments$")
    public void aliceUploadsOversizedFile() throws Exception {
        String token = state.getAccessToken();
        String expenseId = state.getExpenseId();
        Assertions.assertNotNull(expenseId, "Expense ID must be set");

        byte[] oversizedContent = new byte[MAX_FILE_SIZE + 1024];
        ServiceResponse response = AppFactory.getService()
                .uploadAttachment(token, expenseId, "big.jpg", "image/jpeg", oversizedContent);
        state.setLastResponse(response);
    }

    @When("^alice sends GET /api/v1/expenses/\\{expenseId\\}/attachments$")
    public void aliceSendsGetAttachments() throws Exception {
        String token = state.getAccessToken();
        String expenseId = state.getExpenseId();
        Assertions.assertNotNull(expenseId, "Expense ID must be set");

        ServiceResponse response = AppFactory.getService()
                .listAttachments(token, expenseId);
        state.setLastResponse(response);
    }

    @When("^alice sends GET /api/v1/expenses/\\{bobExpenseId\\}/attachments$")
    public void aliceSendsGetBobsAttachments() throws Exception {
        String token = state.getAccessToken();
        String expenseId = state.getBobExpenseId();
        Assertions.assertNotNull(expenseId, "Bob's expense ID must be set");

        ServiceResponse response = AppFactory.getService()
                .listAttachments(token, expenseId);
        state.setLastResponse(response);
    }

    @When("^alice sends DELETE /api/v1/expenses/\\{expenseId\\}/attachments/\\{attachmentId\\}$")
    public void aliceSendsDeleteAttachment() throws Exception {
        String token = state.getAccessToken();
        String expenseId = state.getExpenseId();
        String attachmentId = state.getAttachmentId();
        Assertions.assertNotNull(expenseId, "Expense ID must be set");
        Assertions.assertNotNull(attachmentId, "Attachment ID must be set");

        ServiceResponse response = AppFactory.getService()
                .deleteAttachment(token, expenseId, attachmentId);
        state.setLastResponse(response);
    }

    @When("^alice sends DELETE /api/v1/expenses/\\{bobExpenseId\\}/attachments/\\{attachmentId\\}$")
    public void aliceSendsDeleteBobsAttachment() throws Exception {
        String token = state.getAccessToken();
        String expenseId = state.getBobExpenseId();
        String attachmentId = state.getAttachmentId();
        Assertions.assertNotNull(expenseId, "Bob's expense ID must be set");
        Assertions.assertNotNull(attachmentId, "Attachment ID must be set");

        ServiceResponse response = AppFactory.getService()
                .deleteAttachment(token, expenseId, attachmentId);
        state.setLastResponse(response);
    }

    @When("^alice sends DELETE /api/v1/expenses/\\{expenseId\\}/attachments/\\{randomAttachmentId\\}$")
    public void aliceSendsDeleteNonExistentAttachment() throws Exception {
        String token = state.getAccessToken();
        String expenseId = state.getExpenseId();
        Assertions.assertNotNull(expenseId, "Expense ID must be set");

        ServiceResponse response = AppFactory.getService()
                .deleteAttachment(token, expenseId, "99999");
        state.setLastResponse(response);
    }

    @Given("alice has uploaded file {string} with content type {string} to the entry")
    public void aliceHasUploadedFileToEntry(String filename,
            String contentType) throws Exception {
        aliceUploadsFileToExpense(filename, contentType);
        ServiceResponse resp = state.getLastResponse();
        if (resp != null && resp.statusCode() == 201) {
            JsonObject body = resp.body();
            if (body != null) {
                state.setAttachmentId(body.getString("id"));
            }
        }
    }

    @Given("^bob has created an entry with body \\{ \"amount\": \"([^\"]*)\", \"currency\": \"([^\"]*)\", \"category\": \"([^\"]*)\", \"description\": \"([^\"]*)\", \"date\": \"([^\"]*)\", \"type\": \"([^\"]*)\" \\}$")
    public void bobHasCreatedEntry(String amount, String currency, String category,
            String description, String date, String type) throws Exception {
        String bobToken = state.getBobAccessToken();
        Assertions.assertNotNull(bobToken, "Bob's access token must be set");
        ServiceResponse resp = AppFactory.getService()
                .createExpense(bobToken, amount, currency, category, description, date, type);
        Assertions.assertEquals(201, resp.statusCode(),
                "Expected 201 creating bob's entry but got " + resp.statusCode());
        JsonObject body = resp.body();
        Assertions.assertNotNull(body);
        state.setBobExpenseId(body.getString("id"));
    }

    @Then("the response body should contain {int} items in the {string} array")
    public void responseBodyContainsItemsInArray(int count, String field) {
        ServiceResponse response = state.getLastResponse();
        Assertions.assertNotNull(response);
        JsonObject body = response.body();
        Assertions.assertNotNull(body);
        JsonArray arr = body.getJsonArray(field);
        Assertions.assertNotNull(arr, "Expected '" + field + "' array in response");
        Assertions.assertEquals(count, arr.size(),
                "Expected " + count + " items in '" + field + "' but got " + arr.size());
    }

    @Then("the response body should contain an attachment with {string} equal to {string}")
    public void responseBodyContainsAttachmentWithField(String field, String value) {
        ServiceResponse response = state.getLastResponse();
        Assertions.assertNotNull(response);
        JsonObject body = response.body();
        Assertions.assertNotNull(body);
        JsonArray attachments = body.getJsonArray("attachments");
        Assertions.assertNotNull(attachments, "Expected 'attachments' array");
        boolean found = false;
        for (int i = 0; i < attachments.size(); i++) {
            JsonObject att = attachments.getJsonObject(i);
            if (value.equals(att.getString(field))) {
                found = true;
                break;
            }
        }
        Assertions.assertTrue(found,
                "Expected attachment with '" + field + "' = '" + value + "'");
    }

    private byte[] createSampleContent(String filename) {
        if (filename.endsWith(".jpg") || filename.endsWith(".jpeg")) {
            return new byte[]{(byte) 0xFF, (byte) 0xD8, (byte) 0xFF, (byte) 0xE0};
        } else if (filename.endsWith(".pdf")) {
            return "%PDF-1.4 sample".getBytes();
        } else {
            return "sample content".getBytes();
        }
    }
}
